File: py/aoc2023/day7.py
"""
Day 7: Camel Cards
"""

from collections import Counter

SAMPLE_INPUT = """
32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
"""


# pylint: disable=too-many-return-statements
def _rank(hand: tuple[int]):
    counts = Counter(x for x in hand if x >= 0).most_common(2)
    count0 = counts[0][1] if counts else 0
    count1 = counts[1][1] if len(counts) > 1 else 0
    jokers = sum(x < 0 for x in hand)
    if count0 + jokers >= 5:
        return 6
    if count0 + jokers >= 4:
        return 5
    if count0 + count1 + jokers >= 5:
        return 4
    if count0 + jokers >= 3:
        return 3
    if count0 + count1 + jokers >= 4:
        return 2
    if count0 + jokers >= 2:
        return 1
    return 0


def _solve(cards: str, data: str):
    hands = [
        (_rank(hand := tuple(map(cards.find, words[0]))), hand, int(words[1]))
        for line in data.splitlines()
        if len(words := line.split(maxsplit=1)) == 2
    ]
    return sum((i + 1) * bid for i, (_, _, bid) in enumerate(sorted(hands)))


def part1(data):
    """
    >>> part1(SAMPLE_INPUT)
    6440
    """
    return _solve("23456789TJQKA", data)


def part2(data):
    """
    >>> part2(SAMPLE_INPUT)
    5905
    """
    return _solve("23456789TQKA", data)


parts = (part1, part2)

File: rs/src/day7.rs
use itertools::Itertools;
use std::cmp::Ordering;

fn rank(hand: &[Option<usize>]) -> u8 {
    let counts = hand.iter().filter_map(|x| *x).counts();
    let mut iter = counts.values().sorted().rev();
    let count0 = iter.next().copied().unwrap_or_default();
    let count1 = iter.next().copied().unwrap_or_default();
    let jokers = hand.iter().filter(|c| c.is_none()).count();
    if count0 + jokers >= 5 {
        6
    } else if count0 + jokers >= 4 {
        5
    } else if count0 + count1 + jokers >= 5 {
        4
    } else if count0 + jokers >= 3 {
        3
    } else if count0 + count1 + jokers >= 4 {
        2
    } else if count0 + jokers >= 2 {
        1
    } else {
        0
    }
}

fn solve(cards: &[char], data: &str) -> usize {
    data.lines()
        .filter_map(|line| {
            let (hand, bid) = line.split_once(' ')?;
            let hand = hand
                .chars()
                .map(|c| cards.iter().position(|&d| c == d))
                .collect::<Vec<_>>();
            let bid = bid.parse::<usize>().ok()?;
            Some((hand, bid))
        })
        .sorted_unstable_by(|(a, _), (b, _)| {
            a.iter()
                .zip(b.iter())
                .map(|(a, b)| a.cmp(b))
                .find(|&c| c != Ordering::Equal)
                .unwrap_or_else(|| a.len().cmp(&b.len()))
        })
        .sorted_by_cached_key(|(hand, _)| rank(hand))
        .enumerate()
        .map(|(i, (_, bid))| (i + 1) * bid)
        .sum()
}

pub fn part1(data: &str) -> usize {
    solve(&"23456789TJQKA".chars().collect::<Vec<_>>(), data)
}

pub fn part2(data: &str) -> usize {
    solve(&"23456789TQKA".chars().collect::<Vec<_>>(), data)
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        32T3K 765
        T55J5 684
        KK677 28
        KTJJT 220
        QQQJA 483
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(6440, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(5905, part2(EXAMPLE));
    }
}
